### DISCUSSION TITLE
JSON-RPC over HTTP for R-30iB(+) controllers: comet_rpc

### DISCUSSION THREAD

**Discussion by gavanderhoorn** [gavanderhoorn/dominh#20] [General]
For those of you watching this repository (or those that happen to come across this announcement in the future): [gavanderhoorn/comet_rpc](https://github.com/gavanderhoorn/comet_rpc) provides a bare-minimum wrapper around the `COMET` JSON-RPC interface used by FANUC for their newer CRX robots with R-30iB+ controllers.

See the readme in `gavanderhoorn/comet_rpc` for more information.

Main advantage over custom solutions (implemented using Karel, or using the approach Dominh uses): RPCs return in milliseconds, instead of half a second.

**Comment by rastreus** [comment]
Thank you for bringing this to my attention. Is this the Fanuc RPC protocol (`tpc://:3002`) or something different / newer? (https://github.com/onerobotics/go-fanuc/issues/1#issuecomment-502466816)

The project's README mentions:
> FANUC's web server on R-30iB and R-30iB+ controllers (ie: V8.x and up)

_I know it hasn't been tested_, but could this possibly work with a PaintMate 200iA 5L with a R-30iB controller (V8.30)? I'm currently using a solution similar to [dominh](https://github.com/gavanderhoorn/dominh) (KCL and `.stm` pages).

**Reply by gavanderhoorn** [reply]
Could you please post this over at `gavanderhoorn/comet_rpc`?

I'd like to keep this discussion forum on-topic (ie: for Dominh only).

**Reply by gavanderhoorn** [reply]
But: seeing as it is on-topic here:

> I'm currently using a solution similar to [dominh](https://github.com/gavanderhoorn/dominh) (KCL and `.stm` pages).

could you say a little about what made Dominh not usable for you? Different language requirement (seeing as you commented on Jay's Go library)?

**Reply by rastreus** [reply]
Yes, I can post this to `gavanderhoorn/comet_rpc`: [https://github.com/gavanderhoorn/comet_rpc/discussions/3#discussion-4788879](https://github.com/gavanderhoorn/comet_rpc/discussions/3#discussion-4788879)

Correct. My solution is written in .NET (F#/C#). At the time, Dominh was a good proof-of-concept which affirmed that my solution's direction would work.

**Reply by gavanderhoorn** [reply]
I'm guessing/assuming your implementation is not available somewhere?

**Reply by rastreus** [reply]
You are correct. It is a library component in a commercial solution. It is intentionally basic to meet only the required needs. It simply uses KCL and `.stm` files.

Dominh has more functionality.

**Reply by gavanderhoorn** [reply]
> You are correct. It is a library component in a commercial solution.

Unfortunate.

The world of FANUC software and interfaces would be progressing so much faster if everyone would just collaborate a bit more, but somehow that's very difficult in industrial robotics.
